pkg/parser/expression.go

```diff
@@ -0,0 +1,402 @@
+	unaryExp.Right = p.parseExpr(scope, PREFIX)
+	if unaryExp.Right == nil {
+		return nil // previous error
+	}
+	p.validateUnaryType(unaryExp)
+	return unaryExp
+}
+
+func (p *Parser) parseBinaryExpr(scope *scope, left Node) Node {
+	tok := p.cur
+	expType := left.Type()
+	if isComparisonOp(tok.Type) {
+		expType = BOOL_TYPE
+	}
+	binaryExp := &BinaryExpression{Token: tok, T: expType, Op: op(tok), Left: left}
+	prec := p.curPrecedence()
+	p.advance() // advance past operator
+	binaryExp.Right = p.parseExpr(scope, prec)
+	if binaryExp.Right == nil {
+		return nil // previous error
+	}
+	p.validateBinaryType(binaryExp)
+	return binaryExp
+}
+
+func (p *Parser) parseGroupedExpr(scope *scope) Node {
+	p.advance() // advance past (
+	exp := p.parseTopLevelExpr(scope)
+	if !p.assertToken(lexer.RPAREN) {
+		return nil
+	}
+	p.advance() // advance past )
+	return exp
+}
+
+func (p *Parser) parserIndexExpr(scope *scope, left Node) Node {
+	tok := p.cur // TODO: ensure not prefixed by WS
+	p.advance()  // advance past [
+	leftType := left.Type().Name
+	if leftType != ARRAY && leftType != MAP && leftType != STRING {
+		p.appendErrorForToken("only array, string and map type can be indexed, found "+left.Type().Format(), tok)
+		return nil
+	}
+	index := p.parseTopLevelExpr(scope)
+	if !p.assertToken(lexer.RBRACKET) {
+		return nil
+	}
+	p.advance() // advance past ]
+	if leftType == ARRAY || leftType == STRING {
+		if index.Type() != NUM_TYPE {
+			p.appendError(leftType.String() + " index expects num, found " + index.Type().Format())
+			return nil
+		}
+		return &IndexExpression{Token: tok, Left: left, Index: index, T: left.Type().Sub}
+	}
+	if index.Type() != STRING_TYPE {
+		p.appendError("map index expects string, found " + index.Type().Format())
+		return nil
+	}
+	return &IndexExpression{Token: tok, Left: left, Index: index}
+}
+
+func (p *Parser) parserDotExpr(left Node) Node {
+	tok := p.cur // TODO: ensure not prefixed by WS
+	p.advance()  // advance past .
+	leftType := left.Type().Name
+	if leftType != MAP {
+		p.appendErrorForToken("field access with '.' expects map type, found "+left.Type().Format(), tok)
+		return nil
+	}
+	if p.cur.TokenType() != lexer.IDENT {
+		p.appendErrorForToken("expected map key, found "+p.cur.TokenType().Format(), tok)
+		return nil
+	}
+	expr := &DotExpression{Token: tok, Left: left, T: left.Type().Sub, Key: p.cur.Literal}
+	p.advance() // advance past key IDENT
+	return expr
+}
+
+func isBinaryOp(tt lexer.TokenType) bool {
+	return isComparisonOp(tt) || tt == lexer.PLUS || tt == lexer.MINUS || tt == lexer.SLASH || tt == lexer.ASTERISK || tt == lexer.OR || tt == lexer.AND
+}
+
+func isComparisonOp(tt lexer.TokenType) bool {
+	return tt == lexer.EQ || tt == lexer.NOT_EQ || tt == lexer.LT || tt == lexer.GT || tt == lexer.LTEQ || tt == lexer.GTEQ
+}
+
+func (p *Parser) validateUnaryType(unaryExp *UnaryExpression) {
+	tok := unaryExp.Token
+	op := unaryExp.Op
+	if op != OP_MINUS && op != OP_BANG {
```

Comment:
how about:
```
switch unaryExp.Op {
case OP_MINUS:
    if unaryExp.Right.Type() != NUM_TYPE {
        p.appendError...
    }
case OP_BANG:
    if unaryExp.Right.Type() != BOOL_TYPE {
        p.appendError...
    }
default:
    p.appendError... // invalid unary token
}
```

seems more straightforward to me.

Reply:
ok, agreed that's nicer, done.